solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java

```diff
@@ -60,74 +79,192 @@ public String configSetName(CoreDescriptor cd) {
-    throw new UnsupportedOperationException();
+  public void downloadConfig(String configName, Path dest) throws IOException {
+    Path source = getConfigDir(configName);
+    copyRecursively(source, dest);
+  }
+
+  private void copyRecursively(Path source, Path target) throws IOException {
+    try {
+      Files.walkFileTree(
+          source,
+          new SimpleFileVisitor<Path>() {
+            @Override
+            public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs)
+                throws IOException {
+              Files.createDirectories(target.resolve(source.relativize(dir).toString()));
+              return FileVisitResult.CONTINUE;
+            }
+
+            @Override
+            public FileVisitResult visitFile(Path file, BasicFileAttributes attrs)
+                throws IOException {
+              Files.copy(
+                  file, target.resolve(source.relativize(file).toString()), REPLACE_EXISTING);
+              return FileVisitResult.CONTINUE;
+            }
+          });
+    } catch (NoSuchFileException e) {
+      // do nothing
+    }
   }
 
   @Override
   public List<String> listConfigs() throws IOException {
     try (Stream<Path> configs = Files.list(configSetBase)) {
-      return configs.map(Path::getFileName).map(Path::toString).collect(Collectors.toList());
+      return configs
+          .map(Path::getFileName)
+          .map(Path::toString)
+          .sorted()
+          .collect(Collectors.toList());
     }
   }
 
   @Override
-  public byte[] downloadFileFromConfig(String configName, String filePath) throws IOException {
-    throw new UnsupportedOperationException();
+  public byte[] downloadFileFromConfig(String configName, String fileName) throws IOException {
+    Path filePath = getConfigDir(configName).resolve(fileName);
+    byte[] data = null;
+    try {
+      data = Files.readAllBytes(filePath);
+    } catch (NoSuchFileException e) {
+      // do nothing
+    }
+    return data;
   }
 
   @Override
   public List<String> getAllConfigFiles(String configName) throws IOException {
-    throw new UnsupportedOperationException();
+    Path configDir = getConfigDir(configName);
+    List<String> filePaths = new ArrayList<>();
+    Files.walkFileTree(
+        configDir,
+        new SimpleFileVisitor<Path>() {
+          @Override
+          public FileVisitResult visitFile(Path file, BasicFileAttributes attrs)
+              throws IOException {
+            // don't include hidden (.) files
+            if (!Files.isHidden(file)) {
+              filePaths.add(configDir.relativize(file).toString());
+              return FileVisitResult.CONTINUE;
+            }
+            return FileVisitResult.CONTINUE;
+          }
+
+          @Override
+          public FileVisitResult postVisitDirectory(Path dir, IOException ioException) {
+            if (!dir.getFileName().toString().equals(configName)) {
```

Comment:
if the intermediate directory had a name that coincidentally was the same as that of the configSet, this would not do what we want.  I think you could skip this check and instead check that the relatived-dir string path is not empty?